### DISCUSSION TITLE
tutorials/best_practices/godot_notifications

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#264] [User-contributed notes]
# tutorials/best_practices/godot_notifications

Every Object in Godot implements a_notification method. Its purpose is to allow the Object to respond to a variety of engine-level callbacks that may relate to it. For example, if the engine tells ...

https://docs.godotengine.org/en/stable/tutorials/best_practices/godot_notifications.html

**Comment by RudyFisher7** [comment]
There's a gotcha with object initialization. When setting Resource types in the inspector, and leaving all its @exported Variant data members (for example) at default values, the default values will overwrite values set in its _init() function as if they were set via the inspector. Example below:
(https://docs.godotengine.org/en/stable/tutorials/best_practices/godot_notifications.html#init-vs-initialization-vs-export)

```gdscript
class_name MyResource extends Resource
@export var my_int: int = -1

func _init() -> void:
  if my_int < 0: my_int = 2


class_name MyNode extends Node
@export var my_resource := MyResource.new()
```

If my_resource is set in the inspector, MyResource.my_int will be -1. If my_resource isn't set via inspector, it will be 2.

**Reply by RudyFisher7** [reply]
Just to update, I found a work-around for this. Implement a private initialization() function that you call_deferred() at the end of the _init() function. The deferred function will execute after the @export properties are initialized so are overwritten.

**Reply by RudyFisher7** [reply]
This allows to have subclasses set default values that differ from the super class.

**Reply by OffsetMOSFET** [reply]
About subclass default values, if you assign the default value via a function (``_init_my_int``), you can change the default value in subclasses by reimplementing that function.

**Reply by boblako** [reply]
This makes sense. If you are creating a resource inside Godot, you expect @export var to be what you set in the editor.

If you're creating a resource on the fly, however, it's better to make a custom func constructor and call it with updated data.

I don't understand why I need to initialize the my_int variable twice, first in @export and then in init()

**Reply by RudyFisher7** [reply]
If you're referring to my example code snippet, you actually aren't initializing it twice (at least not manually anyway). The steps are:

Scenerio 1:
1. Set my_resource in the inspector (don't set my_resource.my_int in the inspector, just leave that value be at its defaulted -1 value).
2. Result: Even though my_resource.my_int was never set in the inspector, since my_resource itself was, my_resource.my_int is still -1.

Scenerio 2:
1. Set my_resource via GDScript as a member field (\@export var my_resource := MyResource.new() at the top of the script)). Don't set via inspector (inspector shows it's null but the constructor will be called).
2. Result: my_resource.my_int will be 2.

In both scenerios, my_resource.my_int was never set (its value was left at its default value of -1). But since my_resource was set via inspector in scenerio 1, my_resource.my_int's default value of -1 will be treated as if it were set via the inspector.

This has implications when extending MyResource, say, as a specialization with overridden values. The fact that my_int was being assigned in both _init() and @export was for brevity. In practice, the subclass (a specialization) would be implementing the _init() in this example.

**Reply by boblako** [reply]
@export var works the same way as @ready var. The difference between them is that @export is called after init and before ready.

So if the code execution order looks like this:
1. my_int is set in init
2. my_int is set in @export var
3. my_int is reset in call_deferred(), which is called from init

```gdscript
class_name MyResource extends Resource
@export var my_int: int = -1

func _init() -> void:
  if my_int < 0: my_int = 2
  print('first call')
  my_deferred.call_deferred()

func my_deferred():
  my_int = 3
  print('third call')


class_name MyNode extends Node
@export var my_resource := MyResource.new()

func _ready():
print('second call')
print (my_resource.my_int)
```

**Reply by RudyFisher7** [reply]
Your execution order is correct. But I'm not sure what you're trying to say here. I'm simply pointing out that if an exported Resource is assigned via inspector, all of its exported properties will be treated as if also assigned via inspector, whether they actually were or not.

**Comment by pewcworrell** [comment]
I'm debugging some notifications in Godot 4.3. Here's a helper method if anybody else needs it. Note the numbers at the bottom I couldn't find macros for, but I found in the engine source code and may be subject to change. I don't think my list is comprehensive but it's a good start:
```gdscript
func get_notification_name(what: int) -> String:
	match what:
		NOTIFICATION_POSTINITIALIZE: return "NOTIFICATION_POSTINITIALIZE"
		NOTIFICATION_PREDELETE: return "NOTIFICATION_PREDELETE"
		NOTIFICATION_ENTER_TREE: return "NOTIFICATION_ENTER_TREE"
		NOTIFICATION_EXIT_TREE: return "NOTIFICATION_EXIT_TREE"
		NOTIFICATION_READY: return "NOTIFICATION_READY" 
		NOTIFICATION_PARENTED: return "NOTIFICATION_PARENTED" 
		NOTIFICATION_UNPARENTED: return "NOTIFICATION_UNPARENTED" 
		NOTIFICATION_PAUSED: return "NOTIFICATION_PAUSED" 
		NOTIFICATION_UNPAUSED: return "NOTIFICATION_UNPAUSED" 
		NOTIFICATION_PHYSICS_PROCESS: return "NOTIFICATION_PHYSICS_PROCESS" 
		NOTIFICATION_PROCESS: return "NOTIFICATION_PROCESS" 
		NOTIFICATION_INTERNAL_PROCESS: return "NOTIFICATION_INTERNAL_PROCESS" 
		NOTIFICATION_INTERNAL_PHYSICS_PROCESS: return "NOTIFICATION_INTERNAL_PHYSICS_PROCESS" 
		NOTIFICATION_POST_ENTER_TREE: return "NOTIFICATION_POST_ENTER_TREE" 
		NOTIFICATION_DRAG_BEGIN: return "NOTIFICATION_DRAG_BEGIN" 
		NOTIFICATION_DRAG_END: return "NOTIFICATION_DRAG_END" 
		NOTIFICATION_PATH_RENAMED: return "NOTIFICATION_PATH_RENAMED" 
		NOTIFICATION_TRANSLATION_CHANGED: return "NOTIFICATION_TRANSLATION_CHANGED" 
		NOTIFICATION_EDITOR_PRE_SAVE: return "NOTIFICATION_EDITOR_PRE_SAVE" 
		NOTIFICATION_MOVED_IN_PARENT: return "NOTIFICATION_MOVED_IN_PARENT" 
		NOTIFICATION_ENABLED: return "NOTIFICATION_ENABLED" 
		NOTIFICATION_DISABLED: return "NOTIFICATION_DISABLED" 
		NOTIFICATION_SCENE_INSTANTIATED: return "NOTIFICATION_SCENE_INSTANTIATED" 
		NOTIFICATION_WM_MOUSE_ENTER: return "NOTIFICATION_WM_MOUSE_ENTER"  
		NOTIFICATION_WM_MOUSE_EXIT: return "NOTIFICATION_WM_MOUSE_EXIT"  
		NOTIFICATION_WM_WINDOW_FOCUS_IN: return "NOTIFICATION_WM_WINDOW_FOCUS_IN"
		NOTIFICATION_WM_WINDOW_FOCUS_OUT: return "NOTIFICATION_WM_WINDOW_FOCUS_OUT"  
		NOTIFICATION_WM_CLOSE_REQUEST: return "NOTIFICATION_WM_CLOSE_REQUEST"  
		NOTIFICATION_WM_GO_BACK_REQUEST: return "NOTIFICATION_WM_GO_BACK_REQUEST"  
		NOTIFICATION_WM_SIZE_CHANGED: return "NOTIFICATION_WM_SIZE_CHANGED"  
		NOTIFICATION_WM_DPI_CHANGE: return "NOTIFICATION_WM_DPI_CHANGE"  
		NOTIFICATION_VP_MOUSE_ENTER: return "NOTIFICATION_VP_MOUSE_ENTER"  
		NOTIFICATION_VP_MOUSE_EXIT: return "NOTIFICATION_VP_MOUSE_EXIT"  
		NOTIFICATION_OS_MEMORY_WARNING: return "NOTIFICATION_OS_MEMORY_WARNING"  
		NOTIFICATION_WM_ABOUT: return "NOTIFICATION_WM_ABOUT"  
		NOTIFICATION_CRASH: return "NOTIFICATION_CRASH"  
		NOTIFICATION_OS_IME_UPDATE: return "NOTIFICATION_OS_IME_UPDATE"  
		NOTIFICATION_APPLICATION_RESUMED: return "NOTIFICATION_APPLICATION_RESUMED"  
		NOTIFICATION_APPLICATION_PAUSED: return "NOTIFICATION_APPLICATION_PAUSED"  
		NOTIFICATION_APPLICATION_FOCUS_IN: return "NOTIFICATION_APPLICATION_FOCUS_IN"  
		NOTIFICATION_APPLICATION_FOCUS_OUT: return "NOTIFICATION_APPLICATION_FOCUS_OUT"  
		NOTIFICATION_TEXT_SERVER_CHANGED: return "NOTIFICATION_TEXT_SERVER_CHANGED" 
		NOTIFICATION_EDITOR_PRE_SAVE: return "NOTIFICATION_EDITOR_PRE_SAVE" 
		NOTIFICATION_EDITOR_POST_SAVE: return "NOTIFICATION_EDITOR_POST_SAVE"
		3: return "NOTIFICATION_PREDELETE_CLEANUP"
		30: return "NOTIFICATION_DRAW"
		31: return "NOTIFICATION_VISIBILITY_CHANGED"
		32: return "NOTIFICATION_ENTER_CANVAS"
		33: return "NOTIFICATION_EXIT_CANVAS"
		35: return "NOTIFICATION_LOCAL_TRANSFORM_CHANGED"
		36: return "NOTIFICATION_WORLD_2D_CHANGED"
		41: return "NOTIFICATION_ENTER_WORLD"
		42: return "NOTIFICATION_EXIT_WORLD"
		43: return "NOTIFICATION_VISIBILITY_CHANGED"
		44: return "NOTIFICATION_LOCAL_TRANSFORM_CHANGED"
		50: return "NOTIFICATION_BECAME_CURRENT"
		51: return "NOTIFICATION_LOST_CURRENT"
		10000: return "NOTIFICATION_EDITOR_SETTINGS_CHANGED"
		_: return "Unknown Notification (%d)" % what
```
